File: lib/arpack/base/dsband/test/test.dsband.js
/**
* @license MIT
*
* Copyright (c) 2026 Ricky Reusser.
*
* Derived from ARPACK (arpack-ng 3.9.1), Copyright (c) 1996-2008 Rice
* University. Developed by D.C. Sorensen, R.B. Lehoucq, C. Yang, and
* K. Maschhoff, under the BSD-3-Clause license. See LICENSE.txt in the
* repository root for the full license text and upstream attribution.
*/

/* eslint-disable no-restricted-syntax, stdlib/first-unit-test */

// MODULES //

import test from 'node:test';
import assert from 'node:assert/strict';
import Int32Array from '@stdlib/array/int32/lib/index.js';
import Float64Array from '@stdlib/array/float64/lib/index.js';
import dsband from './../lib/dsband.js';


// FUNCTIONS //

function args( over ) {
	// A minimally-sized valid argument list, with named overrides for the throw tests.
	const ncv = 6;
	const n = 8;
	const lda = 4;
	const o = {
		rvec: true,
		howmny: 'all',
		select: new Int32Array( ncv ),
		d: new Float64Array( ncv ),
		Z: new Float64Array( n * ncv ),
		ldz: n,
		sigma: 0.0,
		N: n,
		AB: new Float64Array( lda * n ),
		MB: new Float64Array( lda * n ),
		lda: lda,
		RFAC: new Float64Array( lda * n ),
		kl: 1,
		ku: 1,
		which: 'LM',
		bmat: 'standard',
		nev: 2,
		tol: 0.0,
		resid: new Float64Array( n ),
		ncv: ncv,
		V: new Float64Array( n * ncv ),
		ldv: n,
		iparam: new Int32Array( 11 ),
		workd: new Float64Array( 3 * n ),
		workl: new Float64Array( ( ncv * ncv ) + ( 8 * ncv ) ),
		lworkl: ( ncv * ncv ) + ( 8 * ncv ),
		iwork: new Int32Array( n ),
		infoIn: 0
	};
	if ( over ) {
		Object.assign( o, over );
	}
	return [ o.rvec, o.howmny, o.select, o.d, o.Z, o.ldz, o.sigma, o.N, o.AB, o.MB, o.lda, o.RFAC, o.kl, o.ku, o.which, o.bmat, o.nev, o.tol, o.resid, o.ncv, o.V, o.ldv, o.iparam, o.workd, o.workl, o.lworkl, o.iwork, o.infoIn ];
}


// TESTS //

test( 'dsband is a function', function t() {
	assert.strictEqual( typeof dsband, 'function', 'is a function' );
});

test( 'dsband has expected arity', function t() {
	assert.strictEqual( dsband.length, 28, 'has expected arity' );
});

test( 'dsband throws TypeError for invalid bmat', function t() {
	assert.throws( function throws() {
		dsband.apply( null, args({ bmat: 'X' }) );
	}, TypeError );
});

test( 'dsband throws TypeError for invalid which', function t() {
	assert.throws( function throws() {
		dsband.apply( null, args({ which: 'ZZ' }) );
	}, TypeError );
});

test( 'dsband throws RangeError for negative N', function t() {
	assert.throws( function throws() {
		dsband.apply( null, args({ N: -1 }) );
	}, RangeError );
});

test( 'dsband main API solves a banded generalized eigenproblem', function t() {
	// K x = lambda M x for the 1-D FEM Laplacian K and consistent mass M.
	const n = 20;
	const nev = 4;
	const ncv = 10;
	const lda = 4;
	const kl = 1;
	const ku = 1;
	const idiag = kl + ku + 1;
	const isup = kl + ku;
	const isub = kl + ku + 2;
	const AB = new Float64Array( lda * n );
	const MB = new Float64Array( lda * n );
	const h = 1.0 / ( n + 1 );
	const r1 = 4.0 / 6.0;
	const r2 = 1.0 / 6.0;
	let j;
	for ( j = 1; j <= n; j++ ) {
		AB[ ( idiag - 1 ) + ( ( j - 1 ) * lda ) ] = 2.0 / h;
		MB[ ( idiag - 1 ) + ( ( j - 1 ) * lda ) ] = r1 * h;
	}
	for ( j = 1; j <= n - 1; j++ ) {
		AB[ ( isup - 1 ) + ( j * lda ) ] = -1.0 / h;
		AB[ ( isub - 1 ) + ( ( j - 1 ) * lda ) ] = -1.0 / h;
		MB[ ( isup - 1 ) + ( j * lda ) ] = r2 * h;
		MB[ ( isub - 1 ) + ( ( j - 1 ) * lda ) ] = r2 * h;
	}

	const RFAC = new Float64Array( lda * n );
	const V = new Float64Array( n * ncv );
	const d = new Float64Array( ncv );
	const resid = new Float64Array( n );
	let r;
	for ( r = 0; r < n; r++ ) {
		resid[ r ] = 1.0 + ( 0.1 * ( r + 1 ) );
	}
	const workd = new Float64Array( 3 * n );
	const lworkl = ( ncv * ncv ) + ( 8 * ncv );
	const workl = new Float64Array( lworkl );
	const iparam = new Int32Array( 11 );
	iparam[ 2 ] = 300;
	iparam[ 6 ] = 3; // shift-invert mode
	const iwork = new Int32Array( n );
	const select = new Int32Array( ncv );

	const info = dsband( true, 'all', select, d, V, n, 0.0, n, AB, MB, lda, RFAC, kl, ku, 'LM', 'generalized', nev, 0.0, resid, ncv, V, n, iparam, workd, workl, lworkl, iwork, 1 );

	assert.strictEqual( info, 0, 'info is 0' );
	assert.strictEqual( iparam[ 4 ], nev, 'all requested eigenvalues converged' );

	// Smallest generalized eigenvalue approximates the fundamental (k*pi)^2:
	assert.ok( Math.abs( d[ 0 ] - ( Math.PI * Math.PI ) ) < 0.05, 'smallest eigenvalue ' + d[ 0 ] + ' near pi^2' );

	// Verify each pair: || K z - d M z || is small (K, M as dense tridiagonals).
	let maxRes = 0.0;
	let c;
	for ( c = 0; c < nev; c++ ) {
		for ( r = 0; r < n; r++ ) {
			let kz = ( 2.0 / h ) * V[ ( c * n ) + r ];
			let mz = ( r1 * h ) * V[ ( c * n ) + r ];
			if ( r > 0 ) {
				kz += ( -1.0 / h ) * V[ ( c * n ) + r - 1 ];
				mz += ( r2 * h ) * V[ ( c * n ) + r - 1 ];
			}
			if ( r < n - 1 ) {
				kz += ( -1.0 / h ) * V[ ( c * n ) + r + 1 ];
				mz += ( r2 * h ) * V[ ( c * n ) + r + 1 ];
			}
			maxRes = Math.max( maxRes, Math.abs( kz - ( d[ c ] * mz ) ) );
		}
	}
	assert.ok( maxRes < 1e-8, 'residual || K z - d M z || is small (' + maxRes + ')' );
});

File: test/fortran/test_dsband.f90
program test_dsband
  use test_utils
  implicit none
  integer, parameter :: MAXN = 24, MAXNCV = 12
  integer, parameter :: MAXBDW = 4, LDA = MAXBDW, LDV = MAXN
  double precision :: a(LDA, MAXN), m(LDA, MAXN), rfac(LDA, MAXN)
  double precision :: v(LDV, MAXNCV), resid(MAXN), d(MAXNCV)
  double precision :: workd(3*MAXN), workl(MAXNCV*MAXNCV + 8*MAXNCV)
  integer :: iparam(11), iwork(MAXN)
  logical :: select(MAXNCV)

  ! ------------------------------------------------------------------
  ! Case 1: generalized shift-invert (type 4), mode 3, bmat='G', LM, sigma=0.
  ! K x = lambda M x with 1-D FEM Laplacian K and consistent mass M.
  call run(20, 4, 10, 3, 'G', 'LM', 0.0d0, 'gen_shiftinv_n20')

  ! ------------------------------------------------------------------
  ! Case 2: standard regular (type 1), mode 1, bmat='I', SA.
  call run(16, 3, 8, 1, 'I', 'SA', 0.0d0, 'std_regular_n16')

  ! ------------------------------------------------------------------
  ! Case 3: standard shift-invert (type 2), mode 3, bmat='I', LM, sigma=1.
  call run(16, 3, 8, 3, 'I', 'LM', 1.0d0, 'std_shiftinv_n16')

contains

  ! Build the band matrices, run dsband, and emit the fixture record.
  subroutine run(n, nev, ncv, mode, bmat, which, sigma, name)
    integer, intent(in) :: n, nev, ncv, mode
    character(*), intent(in) :: bmat, which, name
    double precision, intent(in) :: sigma
    integer :: kl, ku, idiag, isup, isub, j, info, ldz, lworkl, nconv
    double precision :: tol, h, r1, r2
    logical :: rvec

    kl = 1
    ku = 1
    idiag = kl + ku + 1
    isup = kl + ku
    isub = kl + ku + 2

    call dlaset('A', LDA, n, 0.0d0, 0.0d0, a, LDA)
    call dlaset('A', LDA, n, 0.0d0, 0.0d0, m, LDA)
    call dlaset('A', LDA, n, 0.0d0, 0.0d0, rfac, LDA)

    h = 1.0d0 / dble(n + 1)
    r1 = 4.0d0 / 6.0d0
    r2 = 1.0d0 / 6.0d0
    do j = 1, n
      a(idiag, j) = 2.0d0 / h
      m(idiag, j) = r1 * h
    end do
    do j = 1, n - 1
      a(isup, j+1) = -1.0d0 / h
      a(isub, j) = -1.0d0 / h
      m(isup, j+1) = r2 * h
      m(isub, j) = r2 * h
    end do

    ldz = LDV
    lworkl = ncv*ncv + 8*ncv
    tol = 0.0d0
    iparam = 0
    iparam(3) = 300
    iparam(7) = mode
    rvec = .true.

    ! Fixed deterministic initial residual (info != 0 => use provided resid).
    do j = 1, n
      resid(j) = 1.0d0 + 0.1d0 * dble(j)
    end do
    info = 1

    call dsband(rvec, 'A', select, d, v, ldz, sigma, n, a, m, LDA, &
                rfac, kl, ku, which, bmat, nev, tol, resid, ncv, v, LDV, &
                iparam, workd, workl, lworkl, iwork, info)

    nconv = iparam(5)
    call begin_test(name)
    call print_int('n', n)
    call print_int('nev', nev)
    call print_int('nconv', nconv)
    call print_int('mxiter', iparam(3))
    call print_int('info', info)
    call print_array('d', d, nev)
    call print_matrix('z', v, LDV, n, nconv)
    call end_test()
  end subroutine run

end program test_dsband
